### PR TITLE
Feature: PDF Generation

### PR DESCRIPTION
**Pull request by PrayatOsm**
# Changes
1. Added a new namespace with name `Pdf`. The purpose of the namespace is to keep all the classes related to PDF generation operations.
2. Added the following models.
  - `ContentMetaData`
  - `DocumentData`
3. Added a static class `PdfDocumentGenerator`. This static class is exposed as a helper to generate a PDF document.

### CODE REVIEW

File: Pdf/PdfDocumentGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+     
+    public class PdfDocumentGenerator
+    {
+        public static void generatePdfByTemplate(string templatePath, List<ContentMetaData> metaDataList, string outputFilePath)
+        {
+            try 
+            { 
+                if(!File.Exists(templatePath))
+                {
+                    throw new FilePathDoesNotExist("The file path you provided is not Valid");
```
Comment (sameer-s-b) [inline Pdf/PdfDocumentGenerator.cs:26]: Throw an exception object of the Exception class. Remove the custom Exception class.

File: Pdf/PdfDocumentGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    {
+        public FilePathDoesNotExist(string msg) : base(msg)
+        {
+
+        }
+    }
+     
+    public class PdfDocumentGenerator
+    {
+        public static void generatePdfByTemplate(string templatePath, List<ContentMetaData> metaDataList, string outputFilePath)
```
Comment (sameer-s-b) [inline Pdf/PdfDocumentGenerator.cs:20]: `generatePdfByTemplate` -> `GeneratePdfByTemplate`
The convention followed to name the class methods is Pascal case.

File: Pdf/PdfDocumentGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            try 
+            { 
+                if(!File.Exists(templatePath))
+                {
+                    throw new FilePathDoesNotExist("The file path you provided is not Valid"); 
+                }
+                
+                string modifiedHtmlFilePath =  ReplaceFileElementsWithMetaData(templatePath, metaDataList);
+                
+                ConversionCmd(modifiedHtmlFilePath, outputFilePath);
```
Comment (sameer-s-b) [inline Pdf/PdfDocumentGenerator.cs:31]: Rename the method name from `ConversionCmd` to `ConvertHtmlToPdf`

File: Pdf/PdfDocumentGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+            } 
+            catch(Exception e)
+            {
+                throw;
```
Comment (sameer-s-b) [inline Pdf/PdfDocumentGenerator.cs:36]: Use the exception `e`. Throw the exception.

File: Pdf/PdfDocumentGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            }
+        }
+
+        private static string ReplaceFileElementsWithMetaData(string templatePath, List<ContentMetaData> metaDataList)
+        {
+            string htmlContent = File.ReadAllText(templatePath);
+            foreach (ContentMetaData metaData in metaDataList)
+            {
+                htmlContent = htmlContent.Replace("{" + metaData.Placeholder + "}", metaData.Content);
```
Comment (sameer-s-b) [inline Pdf/PdfDocumentGenerator.cs:45]: Use format string to build the string

```
// Old
htmlContent = htmlContent.Replace("{" + metaData.Placeholder + "}", metaData.Content);

// New
htmlContent = htmlContent.Replace($"{{{metaData.Placeholder}}}", metaData.Content);
```

File: Pdf/PdfDocumentGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            string wkHtmlToPdfPath = "cmd.exe";
+
+            string arguments = $"/C Tools\\wkhtmltopdf.exe \"{modifiedHtmlFilePath}\" \"{outputFilePath}\"";
+
+            ProcessStartInfo psi = new ProcessStartInfo
+            {
+                FileName = wkHtmlToPdfPath, // Path to the executable file of wkhtmlpdfpath.
+                Arguments = arguments, // Path to modified template and output location.
+                RedirectStandardOutput = true, // To capture output.
+                RedirectStandardError = true, // To capture error messages.
```
Comment (sameer-s-b) [inline Pdf/PdfDocumentGenerator.cs:75]: Remove the comments. These are not required. Use comments when you need to explain complicated bits of code in simple words.

**Review by sameer-s-b — CHANGES_REQUESTED** [review]

File: Pdf/PdfDocumentGenerator.cs (hunk 2 above)
Comment (PrayatOsm) [reply Pdf/PdfDocumentGenerator.cs:20]: Added pascal case to the class name

File: Pdf/PdfDocumentGenerator.cs (hunk 3 above)
Comment (PrayatOsm) [reply Pdf/PdfDocumentGenerator.cs:31]: renamed class name to ConvertHtmlToPdf
private static void ConvertHtmlToPdf(string modifiedHtmlFilePath, string outputFilePath)

File: Pdf/PdfDocumentGenerator.cs (hunk 4 above)
Comment (PrayatOsm) [reply Pdf/PdfDocumentGenerator.cs:36]: e was unused. throwing it as exception in next line
catch(Exception e)
{
    throw e;
}

File: Pdf/PdfDocumentGenerator.cs (hunk 5 above)
Comment (PrayatOsm) [reply Pdf/PdfDocumentGenerator.cs:45]: using string interpolation in the updated code
foreach (ContentMetaData metaData in metaDataList)
{
    htmlContent = htmlContent.Replace($"{{{metaData.Placeholder}}}", metaData.Content);
}

File: Pdf/PdfDocumentGenerator.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+            }
+        }
+
+        private static string ReplaceFileElementsWithMetaData(string templatePath, List<ContentMetaData> metaDataList)
+        {
+            string htmlContent = File.ReadAllText(templatePath);
+            foreach (ContentMetaData metaData in metaDataList)
+            {
+                htmlContent = htmlContent.Replace("{" + metaData.Placeholder + "}", metaData.Content);
+
```
[the remark this answers is not in the document]
Comment (PrayatOsm) [reply Pdf/PdfDocumentGenerator.cs:46]: removed all unnecessary blank lines in code

File: Pdf/PdfDocumentGenerator.cs (hunk 6 above)
Comment (PrayatOsm) [reply Pdf/PdfDocumentGenerator.cs:75]: removed all comments from process and processStartinfo class

**Review by sameer-s-b — APPROVED** [review]

**Pull request merged**